CMakeLists.txt

```diff
@@ -959,42 +959,58 @@ if(NOT MSVC)
     set(CXX_FLAG -std=c++11)
   else()
 
-    # Try -std=c++14 first
-    if(HPX_WITH_CXX14 OR NOT (("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
-                              AND (CMAKE_CXX_COMPILER_VERSION VERSION_LESS 17)))
-      check_cxx_compiler_flag(-std=c++14 HPX_WITH_CXX14)
-    endif()
+    # Try -std=c++17 first
+    check_cxx_compiler_flag(-std=c++17 HPX_WITH_CXX17)
 
-    if(HPX_WITH_CXX14)
-      set(CXX_FLAG -std=c++14)
-      # The Intel compiler doesn't appear to have a fully functional
-      # implementation of C++14 constexpr. It's fine with our C++14 constexpr
-      # usage in HPX but chokes on Boost.
-      # FIXME: This should be replaced with a version-based check in the future
-      # when the Intel compiler is able to build Boost with -std=c++14.
-      if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Intel")
-        add_definitions(-DBOOST_NO_CXX14_CONSTEXPR)
-      endif()
+    if(HPX_WITH_CXX17)
+      set(CXX_FLAG -std=c++17)
```

Comment:
Since Visual Studio 2015 the Microsoft compiler require a switch for later standard versions too `/std:c++14` and ` /std:c++latest` (see [here](https://blogs.msdn.microsoft.com/vcblog/2016/06/07/standards-version-switches-in-the-compiler/)).

A (maybe better) build system driven way of depending on compiler features is the cmake [`target_compile_features`](https://cmake.org/cmake/help/v3.8/manual/cmake-compile-features.7.html) property, however this feature was introduced in cmake version 3.1.